Prow: adding CLA plugin to Prow

foxish opened the pull request:
Fixes https://github.com/kubernetes/contrib/issues/1834 and will let us turn on the CNCF CLA for all repositories.

/cc @goltermann @bgrant0607 @kubernetes/contributor-experience

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/test-infra/1054)
<!-- Reviewable:end -->

spxtr commented on prow/github/client.go:
```diff
@@ -268,3 +268,25 @@ func (c *Client) CloseIssue(owner, repo string, number int) error {
+
+// FindIssues uses the github search API to find issues which match a particular query.
+func (c *Client) FindIssues(query string) ([]Issue, error) {
+	resp, err := c.request(http.MethodGet, fmt.Sprintf("%s/search/issues?q=%s", c.base, query), nil)
```
Should probably use `net/url.QueryEscape`.

spxtr replied:
Err, maybe not, actually.

spxtr commented on prow/cmd/hook/events.go:
```diff
@@ -81,3 +88,19 @@ func (ea *EventAgent) handleIssueCommentEvent(ic github.IssueCommentEvent) {
+	})
+	l.Infof("Status description %s.", se.Description)
+	for p, h := range ea.Plugins.StatusEventHandlers(se.Repo.FullName) {
+		if err := h(ea.Plugins, se); err != nil {
```
Pass in a `plugins.PluginClient` with the logger `l` as in the other two handlers.

spxtr commented on prow/github/client.go:
```diff
@@ -268,3 +268,25 @@ func (c *Client) CloseIssue(owner, repo string, number int) error {
 }
+
+// FindIssues uses the github search API to find issues which match a particular query.
+func (c *Client) FindIssues(query string) ([]Issue, error) {
```
Add a simple test for this function.

We should consider accepting a `map[string][]string` (or `net/url.Values`) and using `net/url` to make the actual URL, but that can happen later.

spxtr commented on prow/github/types.go:
```diff
@@ -118,3 +118,21 @@ type IssueComment struct {
 }
+
+type StatusEvent struct {
+	SHA         *string `json:"sha,omitempty"`
```
I've been using raw values rather than pointers elsewhere in this file. For consistency's sake, can we keep it that way for now? I understand the reason to use pointers, but they shouldn't really be necessary for this.

Instead of `if se.State == nil || se.Context == nil {` you can do `if se.State == "" || se.Context == "" {` for the same effect, and you don't have to worry about nil dereferences. For instance, you later dereference `se.SHA` without checking that it's not nil. In the strange case that github gives us a payload without a sha, this code crashes the whole ship, whereas not using a pointer here just makes some later function return an error that we log out.

spxtr commented on prow/plugins/cla/cla.go:
```diff
@@ -0,0 +1,110 @@
+	FindIssues(query string) ([]github.Issue, error)
+}
+
+func handleStatusEvent(pa *plugins.PluginAgent, se github.StatusEvent) error {
```
`plugins.PluginClient`

spxtr commented on prow/plugins/util/util.go:
```diff
@@ -0,0 +1,28 @@
+
+import "k8s.io/test-infra/prow/github"
+
+func IssueHasLabel(i github.Issue, labelToFind string) bool {
```
This should go next to `IsAssignee` in `github/types.go`, rather than in a util package.

spxtr commented on prow/plugins/plugins.go:
```diff
@@ -57,6 +58,13 @@ type PluginClient struct {
 	Logger       *logrus.Entry
 }
 
+type StatusEventHandler func(*PluginAgent, github.StatusEvent) error
```
`PluginClient`

spxtr commented on prow/plugins/cla/cla.go:
```diff
@@ -0,0 +1,110 @@
+
+		pr, err := gc.GetPullRequest(org, repo, issue.Number)
+		if err != nil {
+			log.Warningf("Unable to fetch PR-%d from %s/%s.", issue.Number, org, repo)
```
Include the error, `log.WithError(err).Warningf("Unable to fetch PR-%d from %s/%s.", issue.Number, org, repo)`

spxtr commented on prow/plugins/cla/cla.go:
```diff
@@ -0,0 +1,110 @@
+	return handle(pa.GitHubClient, pa.Logger, se)
+}
+
+func handle(gc gitHubClient, log *logrus.Entry, se github.StatusEvent) error {
```
Include a comment describing what's going on with the logic in here, because it's not trivial.

spxtr commented on prow/plugins/cla/cla.go:
```diff
@@ -0,0 +1,110 @@
+
+func handle(gc gitHubClient, log *logrus.Entry, se github.StatusEvent) error {
+	if se.State == nil || se.Context == nil {
+		return fmt.Errorf("Invalid status event delivered with empty state/context")
```
nit: errors should start with lowercase letters

foxish commented:
@spxtr Added tests and addressed comments. PTAL.

spxtr commented on prow/github/client_test.go:
```diff
@@ -266,3 +266,41 @@ func TestCloseIssue(t *testing.T) {
+		t.Errorf("Unexpected number of results: %v", len(result))
+	}
+	if result[0].Number != issueNum {
+		t.Errorf("Expected issue author %+v, got %+v", issueNum, result[0].Number)
```
nit: number.

spxtr commented on prow/plugins/cla/cla_test.go:
```diff
@@ -15,3 +15,157 @@ limitations under the License.
+
+	"github.com/Sirupsen/logrus"
+
+	"fmt"
```
nit: put "fmt" by "testing"

spxtr commented on prow/plugins/cla/cla_test.go:
```diff
@@ -15,3 +15,157 @@ limitations under the License.
+			continue
+		}
+
+		if !reflect.DeepEqual(fc.LabelsAdded, tc.addedLabels) {
```
I'm very wary of reflect. I think you'll only be adding/removing either 0 or 1 labels at a time, so it makes some sense to check that explicitly, then explicitly check that the label that was added/removed is the right one.

Not a big deal though.

spxtr commented on prow/github/client.go:
```diff
@@ -268,3 +268,25 @@ func (c *Client) CloseIssue(owner, repo string, number int) error {
 	return nil
 }
+
+// FindIssues uses the github search API to find issues which match a particular query.
```
Add a TODO saying that we should accept `map[string][]string` and use `net/url` properly.

spxtr commented on prow/plugins/plugins.go:
```diff
@@ -133,3 +141,17 @@ func (pa *PluginAgent) PullRequestHandlers(repo string) map[string]PullRequestHa
 	return hs
 }
+
+// IssueCommentHandlers returns a map of plugin names to handlers for the repo.
```
nit: comment name is wrong

spxtr commented on prow/plugins/cla/cla.go:
```diff
@@ -0,0 +1,119 @@
+			continue
+		}
+
+		// if we end up here, the status is a failure/error.
```
nit: capitalize comments

spxtr commented on prow/plugins/cla/cla.go:
```diff
@@ -0,0 +1,119 @@
+		}
+
+		// if we end up here, the status is a failure/error.
+		// TODO: add a comment which explains what happened and how to rectify it.
```
Add your name to the TODO.

foxish commented:
Thanks @spxtr. Addressed nits and bumped up the version of the hook deployment/makefile. 
We can test it out on kubernetes/pr-bot tomorrow. Then we can turn off the mungebot's cla functionality which is active currently on contrib and replace it.
We also need a way to whitelist an entire org using the plugins.yaml file I think.

foxish commented:
@sarahnovotny, any objections to turning off the google cla bot and switch to the CNCF one completely later this week?

spxtr commented:
> We also need a way to whitelist an entire org using the `plugins.yaml` file I think.

Yep. Right now it's only whitelisting individual repos. I'm thinking either allowing regular expressions like `kubernetes/.*` in `plugins.yaml`, or else somehow specifying that some plugins are opt-out rather than opt-in. Any suggestions?

foxish commented:
@spxtr, I think either the regular expression, or just `org-name` instead of `org-name/repo` in plugins.yaml are good ways to denote it. We may not need opt-out, since at most we'd have 5-6 orgs this tool runs on, and we should be able to specify that.

spxtr commented:
/lgtm

sarahnovotny commented:
@foxish -- please send notice to kubernetes-dev@ that we're planning to gate on the CNCF CLA beginning on $date.   I have no objections.

spxtr commented:
Just `org-name` sgtm. Want to make the change or should I? While we're at it, that package could sure use some testing...
